File: packages/kotlin/cnos-kotlin-aws/src/main/kotlin/ai/kitsy/cnos/vault/aws/AwsSecretsManagerProvider.kt
package ai.kitsy.cnos.vault.aws

import ai.kitsy.cnos.SecretVaultProvider
import ai.kitsy.cnos.SecretVaultProviderFactory
import ai.kitsy.cnos.VaultAuthConfig
import ai.kitsy.cnos.VaultDefinition
import software.amazon.awssdk.regions.Region
import software.amazon.awssdk.services.secretsmanager.SecretsManagerClient
import software.amazon.awssdk.services.secretsmanager.model.GetSecretValueRequest

class AwsSecretsManagerProvider(private val vaultId: String, private val def: VaultDefinition) : SecretVaultProvider {
    private var client: SecretsManagerClient? = null

    override fun authenticate(auth: VaultAuthConfig) {
        val region = def.mapping["region"] ?: System.getenv("AWS_DEFAULT_REGION") ?: "us-east-1"
        client = SecretsManagerClient.builder().region(Region.of(region)).build()
    }

    override fun batchGet(refs: List<String>): Map<String, Any> {
        val c = client ?: return emptyMap()
        val out = mutableMapOf<String, Any>()
        for (ref in refs) {
            try {
                val resp = c.getSecretValue(GetSecretValueRequest.builder().secretId(ref).build())
                val value = resp.secretString() ?: continue
                out[ref] = value
            } catch (_: Exception) {}
        }
        return out
    }

    override fun get(ref: String): Any? {
        return batchGet(listOf(ref))[ref]
    }

    companion object {
        fun factory(): SecretVaultProviderFactory = SecretVaultProviderFactory("aws") { vaultId, def ->
            AwsSecretsManagerProvider(vaultId, def)
        }
    }
}

File: packages/php/cnos-aws/src/AwsSecretsManagerProvider.php
<?php

declare(strict_types=1);

namespace Kitsy\Cnos\Aws;

use Kitsy\Cnos\CnosError;
use Kitsy\Cnos\SecretVaultProvider;
use Kitsy\Cnos\SecretVaultProviderFactory;
use Kitsy\Cnos\VaultAuthConfig;
use Kitsy\Cnos\VaultDefinition;

class AwsSecretsManagerProvider implements SecretVaultProvider
{
    private readonly array $config;
    private ?object $client;

    public function __construct(
        private readonly string $vaultId,
        VaultDefinition $definition,
        ?object $client = null,
    ) {
        $cfg          = $definition->auth->config;
        $this->config = [
            'region'       => (string) ($cfg['region'] ?? ''),
            'endpoint'     => (string) ($cfg['endpoint'] ?? ''),
            'versionId'    => (string) ($cfg['versionId'] ?? $cfg['version'] ?? ''),
            'versionStage' => (string) ($cfg['versionStage'] ?? ''),
        ];
        $this->client = $client;
    }

    public function authenticate(VaultAuthConfig $auth): void
    {
        if ($this->client !== null) return;
        $this->client = $this->buildClient();
    }

    private function buildClient(): object
    {
        if (!class_exists(\Aws\SecretsManager\SecretsManagerClient::class)) {
            throw new CnosError('cnos-aws: aws/aws-sdk-php is required. Run: composer require aws/aws-sdk-php');
        }
        $args = ['version' => 'latest', 'region' => $this->config['region'] ?: 'us-east-1'];
        if ($this->config['endpoint'] !== '') {
            $args['endpoint'] = $this->config['endpoint'];
        }
        return new \Aws\SecretsManager\SecretsManagerClient($args);
    }

    public function batchGet(array $refs): array
    {
        $result = [];
        foreach ($refs as $ref) {
            $val = $this->getOne((string) $ref);
            if ($val !== null) {
                $result[$ref] = $val;
            }
        }
        return $result;
    }

    public function get(string $ref): mixed
    {
        return $this->getOne($ref);
    }

    private function getOne(string $ref): ?string
    {
        try {
            $args = ['SecretId' => $ref];
            if ($this->config['versionId'] !== '')    $args['VersionId']    = $this->config['versionId'];
            if ($this->config['versionStage'] !== '') $args['VersionStage'] = $this->config['versionStage'];

            $resp = $this->client->getSecretValue($args);
            return isset($resp['SecretString']) ? (string) $resp['SecretString'] : null;
        } catch (\Aws\Exception\AwsException $e) {
            if ($e->getAwsErrorCode() === 'ResourceNotFoundException') return null;
            throw new CnosError("cnos-aws: {$e->getMessage()}", $e);
        }
    }

    public static function factory(?object $client = null): SecretVaultProviderFactory
    {
        return new SecretVaultProviderFactory(
            provider: 'aws-secrets-manager',
            create: static fn(string $id, VaultDefinition $def): self => new self($id, $def, $client),
        );
    }
}
